lib/instrumentation/modules/kafkajs.js

```diff
@@ -0,0 +1,472 @@
+   */
+  function wrapProducerSend(origSend) {
+    return async function (record) {
+      const { topic } = record;
+      let span;
+
+      if (!shouldIgnoreTopic(topic, config)) {
+        span = ins.createSpan(
+          `${NAME} send to ${topic}`,
+          TYPE,
+          SUBTYPE,
+          'send',
+          { exitSpan: true },
+        );
+      }
+
+      // W3C trace-context propagation.
+      const runContext = ins.currRunContext();
+      const parentSpan =
+        span || runContext.currSpan() || runContext.currTransaction();
+
+      if (parentSpan) {
+        record.messages.forEach((msg) => {
+          const newHeaders = Object.assign({}, msg.headers);
+          parentSpan.propagateTraceContextHeaders(
+            newHeaders,
+            function (carrier, name, value) {
+              if (name.startsWith('elastic-')) {
+                return;
+              }
+              carrier[name] = value;
+            },
+          );
+          msg.headers = newHeaders;
+        });
+      }
+
+      if (!span) {
+        return origSend.apply(this, arguments);
+      }
+
+      // We do not add headers or body because:
+      // - `record.messages` is a list
+      // - spec says is for transactions (https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#transaction-context-fields)
+      span.setMessageContext({ queue: { name: topic } });
+
+      const service = {
+        resource: `${SUBTYPE}/${topic}`,
+        type: SUBTYPE,
+        name: topic,
+      };
+
+      span._setDestinationContext({ service });
+
+      let result, err;
+      try {
+        result = await origSend.apply(this, arguments);
+      } catch (ex) {
+        // Save the error for use in `finally` below, but re-throw it to
+        // not impact code flow.
+        err = ex;
+        throw ex;
+      } finally {
+        span.setOutcome(
+          err ? constants.OUTCOME_FAILURE : constants.OUTCOME_SUCCESS,
+        );
+        span.end();
+      }
+
+      return result;
+    };
+  }
+
+  /**
+   * Returns the patched version of `Producer.sendBatch` which
+   * - creates an exit span for the operation
+   * - propagates trace context via message headers
+   *
+   * @param {Producer['sendBatch']} origSendBatch
+   * @returns {Producer['sendBatch']}
+   */
+  function wrapProducerSendBatch(origSendBatch) {
+    return async function (batch) {
+      let span;
+      let shouldIgnoreBatch = true;
+      const messages = batch.topicMessages || [];
+      const topics = new Set();
+
+      // Remove possible topic duplications
+      for (const msg of messages) {
+        topics.add(msg.topic);
+      }
+
+      // Check if the batch needs to be ignored
+      for (const t of topics) {
+        if (!shouldIgnoreTopic(t, config)) {
+          shouldIgnoreBatch = false;
+          break;
+        }
+      }
+
+      if (!shouldIgnoreBatch) {
+        span = ins.createSpan(
+          `${NAME} send messages batch`,
```

Comment:
I guess this doesn't conform to https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#naming
However, that spec doesn't cover the batch case.  This is basically the same discussion as https://github.com/elastic/apm-agent-nodejs/pull/3786/files#r1488754514 below.

Reply:
note to self: As agreed, use the following rule for naming batches `Kafaka SEND to A,B,C,D...` where `A,B,C,D...` are the topics sorted and truncated at a certain length

Reply:
... removing ignored topic names from the span/transaction name.

Reply:
when brought to the agents meeting there was a consensus to:
- add the topic name if there is only ONE topic not being ignored
- add no topic names in any form if there is more than one topic not being ignored

Reply:
the implementation is done and a specific test is added in the last commit. @trentm could you do a last review (hopefully)?